Feature: terminate specific Java backend headless instances by port

samhitech opened the pull request:
### Terminate specific Java backend

```Python
def terminate_java_instances(debug=False, port: int = None):
    """
    Terminate headless instances of Micro-Manager core started with the Java backend.

    Parameters
    ----------
    debug : bool
        Print debug messages
    port : int
        Port of the server to terminate. If None, all servers will be terminated
    """
    if not is_java_active():
        logger.debug('No Java instances to stop')
        return

    for key in active_java_ports():
        if port and port != key:
            continue

        process = _JAVA_HEADLESS_SUBPROCESSES[key]
        if debug:
            logger.debug(f'Stopping headless process with pid {process.pid}')
        process.terminate()
        server_terminated(key)
        if debug:
            logger.debug(f'Waiting for process with pid {process.pid} to terminate')
        process.wait()  # wait for process to terminate
        if debug:
            logger.debug(f'Process with pid {process.pid} terminated')
        del _JAVA_HEADLESS_SUBPROCESSES[key]
```

Terminate specific Java backend processes by tracking subprocesses using port numbers. This helps manage multiple headless instances.

```Python
_JAVA_HEADLESS_SUBPROCESSES : dict[int, subprocess.Popen] = {}
```

If the `create_core_instance` function is called using a port that is already in use, an exception will be raised.
```Python
if is_java_port_allocated():
    raise Exception(f'Port {port} already in use')
```

Also, I have split the `terminate_core_instances` function into two
functions:
- `terminate_java_instances`
- `terminate_pymmcore_instances`

Added convenience functions: `is_java_active`, `is_java_port_allocated` and `active_java_ports`.

samhitech committed:
Terminate specific Java backend

Terminate specific Java backend by keeping track of the subprocesses by
port number. This is useful managing multiple headless instances.

Also, I have split the `terminate_core_instances` function into two
functions:
- `terminate_java_instances`
- `terminate_pymmcore_instances`

samhitech committed:
Bump version to 0.3.16

samhitech committed:
Bugfix: `create_core_instance` raises `TypeError`

For `config_file=None`, `create_core_instance` raises `TypeError` when
attempting to convert `config_file` from a relative to an absolute path.
This commit fixes the issue by checking if `config_file` is `None`
before checking if it is a relative path.

samhitech commented:
Hi @henrypinkard! It would be great if you could take the time to look into this.

henrypinkard commented:
Looks good. Could you confirm that the pycro-manager tests are still working as expected with these changes? If so, good to merge

samhitech committed:
Bugfixes and Improvements

- Fixed missing port argument in the `is_java_port_allocated` call
  within `create_core_instance` in `mmpycorex/launcher.py`.

- Enhanced `find_existing_mm_install` in `mmpycorex/install.py` to check
  for existing 'Micro-Manager' or 'Micro-Manager-2.0' directories.

- Use `os.path.join` to construct paths in `find_existing_mm_install`.

- Replaced `get_default_install_location` with
  `find_existing_mm_install`  to automatically determine
  `mm_app_path`.

Passes Pycromanger tests except for `test_viewer.py` which fails is some
cases, like:

```
pytest -v pycromanager/test/test_viewer.py::test_timelapse_napari_viewer[java_backend-save_to_disk]
```

samhitech commented:
Most Pycromanger tests pass, except for `test_viewer.py`, which fails in certain cases, such as:
```PowerShell
pytest -v pycromanager/test/test_viewer.py::test_timelapse_napari_viewer[java_backend-save_to_disk]
```

The issue appears to be related to `napari_signaller`, where the `dataset.is_finished()` method is never set in the test case, leading to an infinite loop that causes _pytest_ to freeze at 98%.
```Python
    @thread_worker(connect={'yielded': update_layer})
    def napari_signaller():
        """
        Monitor for signals that Acquisition has a new image ready, and when that happens
        update napari appropriately
        """
        # don't update faster than the display can handle
        min_update_time = 1 / 30
        last_update_time = time.time()
        while True:
            # import pdb
            # pdb.set_trace()
            dataset_writing_complete = dataset.is_finished()

            if dataset_writing_complete:
                import pdb
                pdb.set_trace()

            new_image_ready = dataset.await_new_image(timeout=.25)
            
            if dataset_writing_complete and not new_image_ready:
                # if the dataset is finished and no new image is ready, exit the loop
                break
            elif not new_image_ready:
                continue

            try:
                # if a new image is ready, get it and update the napari viewer
                image = dataset.as_array()
                update_time = time.time()
                yield image
                if update_time - last_update_time < min_update_time:
                    time.sleep(min_update_time - (update_time - last_update_time))
                last_update_time = time.time()
            except:
                pass
```
This issue does not seem related to this PR. I will also add a PR to pycro-manager soon once I review my edits.

Thanks!

henrypinkard commented:
Thanks for checking

henrypinkard commented:
> Most Pycromanger tests pass, except for `test_viewer.py`, which fails in certain cases

I think this is fine. It is currently passing on the main branch (https://github.com/micro-manager/pycro-manager/actions)
